src/vfs2.c

```diff
@@ -404,128 +490,116 @@ static int check_wal_integrity(sqlite3_file *f)
 	return SQLITE_OK;
 }
 
-static void unregister_file(struct vfs2_file *file)
+/* sqlite3_io_methods implementations begin here */
+
+static sqlite3_file *get_orig(struct file *f)
 {
-	pthread_rwlock_wrlock(&file->vfs_data->rwlock);
-	queue *q = queue_head(&file->vfs_data->queue);
-	while (q != &file->vfs_data->queue) {
-		queue *next = q->next;
-		struct vfs2_db_entry *entry =
-		    QUEUE_DATA(q, struct vfs2_db_entry, link);
-		if (entry->db == file) {
-			entry->db = NULL;
-		} else if (entry->wal == file) {
-			entry->wal = NULL;
-		}
-		if (entry->wal == NULL) {
-			sm_move(&entry->wtx_sm, WTX_NOT_OPEN);
-		}
-		if (entry->db == NULL && entry->wal == NULL) {
-			q->prev->next = q->next;
-			q->next->prev = q->prev;
-			sm_fini(&entry->wtx_sm);
-			sqlite3_free(entry->db_name);
-			sqlite3_free(entry);
-		}
-		q = next;
+	if (f->flags & SQLITE_OPEN_WAL) {
+		return f->entry->wal_cur;
+	} else {
+		return f->orig;
 	}
-	pthread_rwlock_unlock(&file->vfs_data->rwlock);
 }
 
-/* sqlite3_io_methods implementations begin here */
+static void maybe_close_entry(struct entry *e)
+{
+	if (e->refcount_main_db > 0 || e->refcount_wal > 0) {
+		return;
+	}
+
+	sqlite3_free(e->main_db_name);
+	sqlite3_free(e->wal_moving_name);
+	sqlite3_free(e->wal_cur_fixed_name);
+	if (e->wal_cur->pMethods != NULL) {
+		e->wal_cur->pMethods->xClose(e->wal_cur);
+	}
+	sqlite3_free(e->wal_cur);
+	sqlite3_free(e->wal_prev_fixed_name);
+	if (e->wal_prev->pMethods != NULL) {
+		e->wal_prev->pMethods->xClose(e->wal_prev);
+	}
+	sqlite3_free(e->wal_prev);
+
+	free_pending_txn(e);
```

Comment:
I don't think it's a correctness issue to discard the transaction here -- by definition it hasn't been committed yet so there's no loss of data (that we promised to keep). I guess the alternative would be to require calling vfs2_abort before vfs2_close so that the caller has to acknowledge that they're throwing away a transaction? If we did that, what would be the best thing to do in vfs2_close if we see that there's a pending (non-aborted) transaction -- assert or return an error? Or just ignore the pending_txn in close and let the wtx invariant catch if one still exists when we sm_move(WTX_CLOSED)?

Reply:
Thanks for the explanation.